airflow/contrib/hooks/livy_hook.py

```diff
@@ -0,0 +1,297 @@
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+"""
+This module contains the Apache Livy hook.
+"""
+
+import re
+from enum import Enum
+import json
+import requests
+
+from airflow.exceptions import AirflowException
+from airflow.hooks.base_hook import BaseHook
+from airflow.utils.log.logging_mixin import LoggingMixin
+
+
+class BatchState(Enum):
+    """
+    Batch session states
+    """
+    NOT_STARTED = 'not_started'
+    STARTING = 'starting'
+    RUNNING = 'running'
+    IDLE = 'idle'
+    BUSY = 'busy'
+    SHUTTING_DOWN = 'shutting_down'
+    ERROR = 'error'
+    DEAD = 'dead'
+    KILLED = 'killed'
+    SUCCESS = 'success'
+
+
+TERMINAL_STATES = {
+    BatchState.SUCCESS,
+    BatchState.DEAD,
+    BatchState.KILLED,
+    BatchState.ERROR,
+}
+
+
+class LivyHook(BaseHook, LoggingMixin):
+    """
+    Hook for Apache Livy through the REST API.
+
+    For more information about the API refer to
+    https://livy.apache.org/docs/latest/rest-api.html
+
+    :param livy_conn_id: reference to a pre-defined Livy Connection.
+    :type livy_conn_id: str
+    """
+    def __init__(self, livy_conn_id='livy_default'):
+        super(LivyHook, self).__init__(livy_conn_id)
+        self._livy_conn_id = livy_conn_id
+        self._build_base_url()
+
+    def _build_base_url(self):
+        """
+        Build connection URL
+        """
+        params = self.get_connection(self._livy_conn_id)
+
+        base_url = params.host
+
+        if not base_url:
+            raise AirflowException("Missing Livy endpoint hostname")
+
+        if '://' not in base_url:
+            base_url = '{}://{}'.format('http', base_url)
+        if not re.search(r':\d+$', base_url):
+            base_url = '{}:{}'.format(base_url, str(params.port or 8998))
+
+        self._base_url = base_url
+
+    def get_conn(self):
+        pass
+
+    def post_batch(self, *args, **kwargs):
+        """
+        Perform request to submit batch
+        """
+
+        batch_submit_body = json.dumps(LivyHook.build_post_batch_body(*args, **kwargs))
+        headers = {'Content-Type': 'application/json'}
+
+        self.log.info("Submitting job {} to {}".format(batch_submit_body, self._base_url))
+        response = requests.post(self._base_url + '/batches', data=batch_submit_body, headers=headers)
```

Comment:
- Add `verify` argument to optionally point to an SSL trust store. Our organization has a valid (purchased) SSL certificate, but an intermediate CA is not in a list of available by default, so we have to point clients to a Trust Store (a modified pem file) that has that intermediate CA added. Otherwise clients fail SSL negotiation. https://github.com/Tagar/livy/blob/master/test_livy.py#L108

- Would need a way to optionally add `auth` argument too. See example here - https://github.com/Tagar/livy/blob/master/test_livy.py#L106 
This is required for Kerberos authentication for example that our organization uses. 

To satisfy both of the above changes, perhaps add additional `requests_args` argument to `post_batch`, or perhaps even better - to the constructor of LivyHook as exact same additional parameters are required for the all other requests to Livy. `get_batch_state`, `delete_batch`, etc

Reply:
These parameters are a great addition, but should come from the `extra` argument in the Connection.

Reply:
(And basing this hook on the HttpHook would mean we _already_ support the verify argument.)

Reply:
Thanks @ashb 
Would we be able to add 
```
 'X-Requested-By': 'livy-client'
```
too that I referenced above? 
Through `extra` argument when LivyHook will be based off of HttpHook?

Reply:
@ashb `HttpHook` doesn't support currently `auth` if I understand that right 
https://github.com/apache/airflow/blob/master/airflow/hooks/http_hook.py#L165

Reply:
If username and password is all that is needed that is already handled: https://github.com/apache/airflow/blob/master/airflow/hooks/http_hook.py#L72-L73

If you need something more you can override `get_conn` in the subclass to set `auth` on the session object https://requests.kennethreitz.org//en/master/api/#requests.Session.auth

```python
    def get_conn(self, headers=None):
        session = super().get_conn(headers)
        session.auth = ...
        return session
```

Reply:
@ashb I was looking for Kerberos authentication so useername/password wouldn't do, and overriding `get_conn` in the subclass is the way to do it - thank you for pointing this out

Reply:
these kind of parameters ideally should be configurable in the connection (eg: `extra = {"headers": {...}, "auth": ..., "verify": ...}`), but since currently `HttpHook` dumps the whole `extra` in the headers (https://github.com/apache/airflow/blob/00ef8cb92857fa2837112b6480a4a7c152073099/airflow/hooks/http_hook.py#L74-L76) that would mean to completely override how `get_conn` interprets this field.

Unless we want to completely change how the `extra` fields is managed by the parent class the only solution is to expose these parameters in the constructor.

Reply:
We can perhaps do that in a follow up PR probably by adding `kerberos` authentication to httpHook itself? Workaround is creating a subclass that overrides `get_conn` as @ashb for suggesting. 

Another option is just to add `auth` of type [HTTPKerberosAuth](https://github.com/Tagar/livy/blob/master/test_livy.py#L106 ) into `conn.extra_dejson` directly ( I haven't tested this, but can do this after this PR is committed)

Reply:
the current HttpHook implementation would try to put the `auth` as a header.
Probably the best solution is to update directly HttpHook, but that would require either adding a new field to the connection or change the `extra` usage (which would break someone's code)